File: app-android/src/main/java/org/mtransit/android/analytics/AnalyticsManager.java
package org.mtransit.android.analytics;

import android.content.Context;
import android.os.Bundle;

import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.Size;

import com.google.firebase.analytics.FirebaseAnalytics;

import org.mtransit.android.commons.MTLog;
import org.mtransit.android.ui.view.common.IActivity;

import java.util.Map;

import javax.inject.Inject;

import dagger.hilt.android.qualifiers.ApplicationContext;

// ENABLE:
// - DEBUG logging:
// adb shell setprop log.tag.FA VERBOSE
// adb shell setprop log.tag.FA-SVC VERBOSE
// - FASTER EVENTS
// adb shell setprop debug.firebase.analytics.app org.mtransit.android
//
// LOGCAT:
// adb logcat -v time -s FA FA-SVC
//
// DISABLE:
// - DEBUG logging:
// adb shell setprop log.tag.FA \"\"
// adb shell setprop log.tag.FA-SVC \"\"
// - FASTER EVENTS
// adb shell setprop debug.firebase.analytics.app .none.
//
// https://firebase.google.com/docs/analytics/debugview
public class AnalyticsManager implements IAnalyticsManager, MTLog.Loggable {

	private static final String LOG_TAG = AnalyticsManager.class.getSimpleName();

	@NonNull
	@Override
	public String getLogTag() {
		return LOG_TAG;
	}

	private static final boolean ANALYTICS_ENABLED = true;
	// private static final boolean ANALYTICS_ENABLED = false; // DEBUG

	private static final boolean DEBUG = false;
	// private static final boolean DEBUG = true; // DEBUG

	@NonNull
	private final FirebaseAnalytics firebaseAnalytics;

	@Inject
	public AnalyticsManager(@NonNull @ApplicationContext Context appContext) {
		if (!ANALYTICS_ENABLED) {
			//noinspection ConstantConditions
			firebaseAnalytics = null;
			return;
		}
		firebaseAnalytics = FirebaseAnalytics.getInstance(appContext);
		if (DEBUG) {
			// DEBUG adb shell setprop debug.firebase.analytics.app org.mtransit.android
			// DEBUG adb shell setprop log.tag.FA VERBOSE
			// DEBUG adb shell setprop log.tag.FA-SVC VERBOSE
			// DEBUG adb logcat -v time -s FA FA-SVC
		}
	}

	@Override
	public void setUserProperty(@NonNull @Size(min = 1L, max = 24L) String name,
								int value) {
		setUserProperty(name, String.valueOf(value));
	}

	@Override
	public void setUserProperty(@NonNull @Size(min = 1L, max = 24L) String name,
								@NonNull String value) {
		if (!ANALYTICS_ENABLED) {
			return;
		}
		try {
			firebaseAnalytics.setUserProperty(name, value);
		} catch (Exception e) {
			MTLog.w(this, e, "Error while tracing user property (%s:%s)", name, value);
		}
	}

	@Override
	public void logEvent(@NonNull @Size(min = 1L, max = 40L) String name) {
		if (!ANALYTICS_ENABLED) {
			return;
		}
		logEvent(name, null);
	}

	@Override
	public void logEvent(@NonNull @Size(min = 1L, max = 40L) String name,
						 @Nullable AnalyticsEventsParamsProvider params) {
		if (!ANALYTICS_ENABLED) {
			return;
		}
		Bundle bundle = null;
		if (params != null) {
			Map<String, Object> paramMap = params.to();
			bundle = new Bundle();
			for (Map.Entry<String, Object> param : paramMap.entrySet()) {
				// Firebase: "String, long and double param types are supported."
				if (param.getValue() instanceof String) {
					bundle.putString(param.getKey(), (String) param.getValue());
				} else if (param.getValue() instanceof Long) {
					bundle.putLong(param.getKey(), (Long) param.getValue());
				} else if (param.getValue() instanceof Double) {
					bundle.putDouble(param.getKey(), (Double) param.getValue());
				} else {
					MTLog.w(this, "Unexpected event parameter type for '" + param.getKey() + "'>'" + param.getValue() + "'!");
				}
			}
		}
		this.firebaseAnalytics.logEvent(name, bundle);
	}

	@MainThread
	@Override
	public void trackScreenView(@NonNull Trackable page) {
		if (!ANALYTICS_ENABLED) {
			return;
		}
		try {
			Bundle bundle = new Bundle();
			bundle.putString(FirebaseAnalytics.Param.SCREEN_NAME, page.getScreenName());
			bundle.putString(FirebaseAnalytics.Param.SCREEN_CLASS, page.getClass().getSimpleName());
			firebaseAnalytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, bundle);
		} catch (Exception e) {
			MTLog.w(this, e, "Error while tracing screen view! (%s)", page);
		}
	}
}

File: app-android/src/main/java/org/mtransit/android/analytics/AnalyticsManager.kt
package org.mtransit.android.analytics

import android.content.Context
import android.os.Bundle
import androidx.annotation.MainThread
import androidx.annotation.Size
import com.google.firebase.analytics.FirebaseAnalytics
import dagger.hilt.android.qualifiers.ApplicationContext
import org.mtransit.android.commons.MTLog
import org.mtransit.android.commons.MTLog.Loggable
import javax.inject.Inject

// ENABLE:
// - DEBUG logging:
// adb shell setprop log.tag.FA VERBOSE
// adb shell setprop log.tag.FA-SVC VERBOSE
// - FASTER EVENTS
// adb shell setprop debug.firebase.analytics.app org.mtransit.android
//
// LOGCAT:
// adb logcat -v time -s FA FA-SVC
//
// DISABLE:
// - DEBUG logging:
// adb shell setprop log.tag.FA \"\"
// adb shell setprop log.tag.FA-SVC \"\"
// - FASTER EVENTS
// adb shell setprop debug.firebase.analytics.app .none.
//
// https://firebase.google.com/docs/analytics/android/events#view-events-in-debug-log
// https://firebase.google.com/docs/analytics/debugview
class AnalyticsManager @Inject internal constructor(
    @ApplicationContext appContext: Context,
) : IAnalyticsManager, Loggable {

    companion object {
        private val LOG_TAG: String = AnalyticsManager::class.java.getSimpleName()

        private const val ANALYTICS_ENABLED = true

        // private static final boolean ANALYTICS_ENABLED = false // DEBUG
        private const val DEBUG = false
    }

    override fun getLogTag() = LOG_TAG

    // private static final boolean DEBUG = true; // DEBUG
    private val firebaseAnalytics: FirebaseAnalytics?

    init {
        if (!ANALYTICS_ENABLED) {
            firebaseAnalytics = null
        } else {
            firebaseAnalytics = FirebaseAnalytics.getInstance(appContext)
            @Suppress("ControlFlowWithEmptyBody")
            if (DEBUG) {
                // DEBUG adb shell setprop debug.firebase.analytics.app org.mtransit.android
                // DEBUG adb shell setprop log.tag.FA VERBOSE
                // DEBUG adb shell setprop log.tag.FA-SVC VERBOSE
                // DEBUG adb logcat -v time -s FA FA-SVC
            }
        }
    }

    override fun setUserProperty(
        @Size(min = 1L, max = 24L) name: String,
        value: Int
    ) {
        setUserProperty(name, value.toString())
    }

    override fun setUserProperty(
        @Size(min = 1L, max = 24L) name: String,
        value: String
    ) {
        if (!ANALYTICS_ENABLED) return
        try {
            firebaseAnalytics?.setUserProperty(name, value)
        } catch (e: Exception) {
            MTLog.w(this, e, "Error while tracing user property ($name:$value)")
        }
    }

    override fun logEvent(@Size(min = 1L, max = 40L) name: String) {
        if (!ANALYTICS_ENABLED) return
        logEvent(name, null)
    }

    override fun logEvent(
        @Size(min = 1L, max = 40L) name: String,
        params: AnalyticsEventsParamsProvider?
    ) {
        if (!ANALYTICS_ENABLED) return
        val bundle = params?.let {
            Bundle().apply {
                for ((key, value) in it.to()) {
                    when (value) {
                        is String -> putString(key, value)
                        is Boolean -> putBoolean(key, value)
                        is Int -> putInt(key, value)

                        else -> MTLog.w(this@AnalyticsManager, "Unexpected event parameter type for '${key}'>'${value}'!")
                    }
                }
            }
        }
        this.firebaseAnalytics?.logEvent(name, bundle)
    }

    @MainThread
    override fun trackScreenView(page: AnalyticsScreen) {
        if (!ANALYTICS_ENABLED) return
        try {
            firebaseAnalytics?.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, Bundle().apply {
                putString(FirebaseAnalytics.Param.SCREEN_NAME, page.screenName)
                putString(FirebaseAnalytics.Param.SCREEN_CLASS, page.screenClass)
            })
        } catch (e: Exception) {
            MTLog.w(this, e, "Error while tracking screen view! ($page)")
        }
    }

    @MainThread
    override fun trackButtonClick(buttonName: String, page: AnalyticsScreen?) {
        if (!ANALYTICS_ENABLED) return
        try {
            firebaseAnalytics?.logEvent(AnalyticsEvents.BUTTON_CLICK, Bundle().apply {
                putString(AnalyticsEvents.Params.BUTTON_NAME, buttonName)
                page?.let {
                    putString(FirebaseAnalytics.Param.SCREEN_NAME, it.screenName)
                    putString(FirebaseAnalytics.Param.SCREEN_CLASS, it.screenClass)
                }
            })
        } catch (e: Exception) {
            MTLog.w(this, e, "Error while tracking button $buttonName click! (page:$page)")
        }
    }
}
